File: all-code/701-800/790numTilings.py
# 有两种形状的瓷砖：一种是 2 x 1 的多米诺形，另一种是形如 "L" 的托米诺形。两种形状都可以旋转。
#
#
#
# 给定整数 n ，返回可以平铺 2 x n 的面板的方法的数量。返回对 109 + 7 取模 的值。
#
# 平铺指的是每个正方形都必须有瓷砖覆盖。两个平铺不同，当且仅当面板上有四个方向上的相邻单元中的两个，使得恰好有一个平铺有一个瓷砖占据两个正方形。
#
#
#
# 示例 1:
#
#
#
# 输入: n = 3
# 输出: 5
# 解释: 五种不同的方法如上所示。
# 示例 2:
#
# 输入: n = 1
# 输出: 1
#
#
# 提示：
#
# 1 <= n <= 1000

class Solution:
    def numTilings(self, n: int) -> int:
        f, g = [0] * n, [0] * n
        MOD = int(1e9) + 7
        f[0], g[0] = 1, 0
        if n > 1: f[1], g[1] = 2, 1
        for i in range(2, n):
            f[i] = f[i - 1] + f[i - 2] + g[i - 1] * 2
            f[i] %= MOD
            g[i] = f[i - 2] + g[i - 1]
            g[i] %= MOD
        return f[-1]



so = Solution()
print(so.numTilings(4))
print(so.numTilings(3))
print(so.numTilings(1))

File: c++/src/1-999/790numTilings.cpp
#include "lc_pub.h"

using namespace std;

class Solution {
public:

int numTilings1(int n) {
    if (n == 1) return 1;
    int MOD = 1'000'000'007;
    vector<vector<long long>> dp(n+1, vector<long long>(n+1, 0LL));  // 第一排长度为i，第二排长度为j的所有可能
    dp[0][0] = 1;
    for (int i=0;i<n+1;i++) {
        for (int j=0;j<n+1;j++) {
            if (i<=0&&j<=0) continue;
            if (i - j >= 2)
                dp[i][j]=dp[i-2][j];
            else if (j - i >= 2)
                dp[i][j]=dp[i][j-2];
            else if (i == j)
                if (i>1)
                    dp[i][j]=dp[i-1][j-1]+dp[i-2][j]+dp[i-2][j-1]+dp[i-1][j-2];
                else
                    dp[i][j]=dp[i-1][j-1];
            else if (i > j){
                if (j > 0)
                    dp[i][j]=dp[i-2][j]+dp[i-2][j-1];
            }
            else
            {
                if (i>0)
                    dp[i][j]=dp[i][j-2]+dp[i-1][j-2];
            }
            dp[i][j]%=MOD;
        }
    }
    return dp[n][n];
    }
int numTilings(int n) {
    if (n == 1) return 1;
    int MOD = 1'000'000'007;
    vector<vector<long long>> dp(n+1, vector<long long>(4, 0LL));  // dp[i][j] 表示处理第i列时，i列的四种情况的所有4种可能
    dp[0][0] = 1;
    for (int i=1;i<n+1;i++) {
        dp[i][0]=(dp[i-1][0]+dp[i-1][1]+dp[i-1][2]+dp[i-1][3])%MOD;  // 两个格子是填满的
        dp[i][1]=(dp[i-1][3]+dp[i-1][2])%MOD;
        dp[i][2]=(dp[i-1][3]+dp[i-1][1])%MOD;
        dp[i][3]=(dp[i-1][0])%MOD;
    }
    return dp[n][0];
}
};

int main()
{
    std::cout<<"test let us start! %s" << __cplusplus <<std::endl;
    vector<int> arrays = {1,1,2};
    Solution so;
    cout << so.numTilings(4) <<endl;
    return 0;
}